### DISCUSSION TITLE
Consider official Emscripten devcontainer feature

### DISCUSSION THREAD

**Discussion by jcbhmr** [emscripten-core/emscripten#18944] [Ideas]
Hello! 👋

I'm interested in seeing if this @emscripten-core organization would be **willing to _consider_** a emscripten-core/devcontainer-features repository to house an official Emscripten [devcontainer feature](https://code.visualstudio.com/blogs/2022/09/15/dev-container-features) for [GitHub Codespaces](https://github.com/features/codespaces) (and other devcontainer environments).

Example:

```jsonc
// .devcontainer/devcontainer.json
{
  "features": {
    "ghcr.io/emscripten-core/devcontainer-features/emsdk": {}
  }
}
```

In the current official [devcontainers/features](https://github.com/devcontainers/features) repository, they are **not accepting new features** and are instead [encouraging users to make their own](https://github.com/devcontainers/feature-starter). I think it would be beneficial to follow in the footsteps of other organizations like [Julia](https://github.com/JuliaLang/devcontainer-features), [R](https://github.com/rocker-org/devcontainer-features), [Mamba](https://github.com/mamba-org/devcontainer-features), and others by creating and maintaining an official Emscripten installer feature.

If this isn't tenable, that's OK! 👍 It's possible (and even probable) that there will be some kind of @devcontainers-community features repository that could house community-driven features. It'd just be nice if it were under the official umbrella.

Stemmed from [devcontainers-community#1](https://github.com/orgs/devcontainers-community/discussions/1#discussioncomment-5278444)

**Comment by sbc100** [comment]
First time I'm learning about devcontainsers.  From the looks of it seems to be a feature of visual studio code.  In that case what has this got to do with Deno?

Emscripten already publishs docker container with each emsdk release here: https://hub.docker.com/r/emscripten/emsdk.   Can this be used as our devcontainer?

**Reply by jcbhmr** [reply]
> First time I'm learning about devcontainsers. From the looks of it seems to be a feature of visual studio code. In that case what has this got to do with Deno?

Yep, it's primarily a feature of VS Code & GitHub Codespaces. With the advent of GitHub Codespaces being free for 60hrs/month, it's a reasonably accessible way to jump in to code with a few clicks.

It's also integrated with some other IDEs:
<img width=500 src="https://user-images.githubusercontent.com/61068799/225132392-0de0fc29-e3da-422c-b6e6-715669d1b7a8.png" />
https://github.blog/2023-02-28-10-things-you-didnt-know-you-could-do-with-github-codespaces/

Sorry, I duplicated this post almost word for word from https://github.com/denoland/deno/discussions/18144. My bad! 🤦‍♂️ I edited it to say "Emscripten".

> Emscripten already publishs docker container with each emsdk release here: https://hub.docker.com/r/emscripten/emsdk. Can this be used as our devcontainer?

Hmmm... 🤔 This _may_ (haven't tested) be usable as the `image` field, but it's not a devcontainer _feature_. A feature is more of an installer script generalized for any base image that gets layered on top of another image. You can find an existing implementation of an emscripten devcontainer feature here: https://github.com/ebaskoro/devcontainer-features/tree/main/src/emscripten

An example usecase would be a Node.js app that uses the base image of mcr.microsoft.com/devcontainers/javascript-node:18 and then has the features of ghcr.io/emscripten-core/devcontainer-features/emsdk and maybe some Java addon with ghcr.io/devcontainers/features/java

The reason that I'm bringing this to your attention is part of an overture to get "official" devcontainer features for tools like Emscripten, Deno, Haskell, Julia, R, etc. since the official devcontainers/features repo doesn't intend to expand beyond their core set of maintained features.

This would be a way to potentially get more beginners (like me) using Emscripten, maybe increase the ease of getting started to "Just click Open in Codespaces to get started!" or something like that. Beginner accessibility is a good thing IMO! ❤

The reason for centralizing such things into the Emscripten organization is to have _multiple people with write access_ to avoid decay. It would also lend some legitimacy to such a feature as being "official" and the defacto standard instead of potentially multiple variations with no official version. So it basically comes down to branding I guess if you really want to reduce it.

/cc @ebaskoro

**Reply by eitsupi** [reply]
> First time I'm learning about devcontainsers. From the looks of it seems to be a feature of visual studio code.

Note that this was originally a closed-source extension for VS Code, but is now open-specified and available through the OSS CLI.
